File: frontend/src/constants/roleType.ts
/**
 * Role Type Constants
 *
 * Unified role type definitions for academic talent identification.
 * This file provides a single source of truth for role type display
 * configuration across the frontend application.
 *
 * Role Types:
 * - professor: 教授/研究员 (h_index >= 25 or high citation count)
 * - student: 在读学生 (works <= 8 and low citations)
 * - graduate: 毕业/早期研究者 (8 < works < 30, transitioning)
 * - unknown: 未知 (insufficient data)
 */

export type RoleTypeValue = 'professor' | 'student' | 'graduate' | 'unknown'

export interface RoleTypeConfig {
  color: string
  text: string
}

/**
 * Role type display configuration map
 */
export const ROLE_TYPE_MAP: Record<RoleTypeValue, RoleTypeConfig> = {
  professor: { color: 'green', text: '教授/研究员' },
  student: { color: 'blue', text: '学生' },
  graduate: { color: 'orange', text: '毕业生' },
  unknown: { color: 'default', text: '未知' },
}

/**
 * Legacy role type mappings for backward compatibility
 * Maps old role types to new unified types
 */
export const LEGACY_ROLE_TYPE_MAP: Record<string, RoleTypeValue> = {
  // Old values that need to be mapped
  graduated: 'graduate',
  teaching_research: 'professor',
  associate_professor: 'professor',
  researcher: 'graduate',
  phd_student: 'student',
  master_student: 'student',
}

/**
 * Get role type configuration with backward compatibility
 *
 * @param roleType - Role type string (may be legacy or current)
 * @returns RoleTypeConfig with color and display text
 */
export function getRoleTypeConfig(roleType: string): RoleTypeConfig {
  // First, check if it's a legacy type that needs mapping
  const normalizedType = LEGACY_ROLE_TYPE_MAP[roleType] || roleType

  // Return config for the normalized type, fallback to unknown
  return ROLE_TYPE_MAP[normalizedType as RoleTypeValue] || ROLE_TYPE_MAP.unknown
}

/**
 * Get display text for a role type
 *
 * @param roleType - Role type string
 * @returns Chinese display text
 */
export function getRoleTypeText(roleType: string): string {
  return getRoleTypeConfig(roleType).text
}

/**
 * Get color for a role type
 *
 * @param roleType - Role type string
 * @returns Ant Design tag color
 */
export function getRoleTypeColor(roleType: string): string {
  return getRoleTypeConfig(roleType).color
}

/**
 * Check if a role type is valid
 *
 * @param roleType - Role type string to validate
 * @returns true if valid role type
 */
export function isValidRoleType(roleType: string): roleType is RoleTypeValue {
  return roleType in ROLE_TYPE_MAP
}

/**
 * Get all role type options for select/ filter components
 *
 * @returns Array of role type options with value and label
 */
export function getRoleTypeOptions(): Array<{ value: RoleTypeValue; label: string }> {
  return [
    { value: 'professor', label: '教授/研究员' },
    { value: 'student', label: '学生' },
    { value: 'graduate', label: '毕业生' },
    { value: 'unknown', label: '未知' },
  ]
}

File: backend/app/constants/role_type.py
"""
Role Type Constants

Unified role type definitions for academic talent identification.
Matches frontend/src/constants/roleType.ts
"""

from enum import Enum


class RoleType(str, Enum):
    """Role type enum for academic talent classification."""

    PROFESSOR = "professor"  # 教授/研究员 (h_index >= 25 or high citation count)
    STUDENT = "student"  # 在读学生 (works <= 8 and low citations)
    GRADUATE = "graduate"  # 毕业/早期研究者 (8 < works < 30, transitioning)
    UNKNOWN = "unknown"  # 未知 (insufficient data)

    @property
    def display_text(self) -> str:
        """Get Chinese display text for the role type."""
        texts = {
            RoleType.PROFESSOR: "教授/研究员",
            RoleType.STUDENT: "学生",
            RoleType.GRADUATE: "毕业生",
            RoleType.UNKNOWN: "未知",
        }
        return texts.get(self, "未知")

    @property
    def color(self) -> str:
        """Get Ant Design tag color for the role type."""
        colors = {
            RoleType.PROFESSOR: "green",
            RoleType.STUDENT: "blue",
            RoleType.GRADUATE: "orange",
            RoleType.UNKNOWN: "default",
        }
        return colors.get(self, "default")


# Legacy role type mappings for backward compatibility
LEGACY_ROLE_TYPE_MAP: dict[str, str] = {
    "graduated": "graduate",
    "teaching_research": "professor",
    "associate_professor": "professor",
    "researcher": "graduate",
    "phd_student": "student",
    "master_student": "student",
}


def normalize_role_type(role_type: str) -> str:
    """
    Normalize role type string with backward compatibility.

    Args:
        role_type: Role type string (may be legacy or current)

    Returns:
        Normalized role type string
    """
    if role_type in LEGACY_ROLE_TYPE_MAP:
        return LEGACY_ROLE_TYPE_MAP[role_type]
    return role_type
